modules/gapi/src/executor/gtbbstreamingexecutor.cpp

```diff
@@ -0,0 +1,1728 @@
+        if (capacity != 0) {
+            ptr->set_capacity(capacity);
+        }
+        q = std::move(ptr);
+    }
+    explicit DataQueue(tag t)
+        : q(new DesyncQueue()) {
+        GAPI_Assert(t == DESYNC);
+    }
+
+    // FIXME: ADE metadata requires types to be copiable
+    std::shared_ptr<Q> q;
+};
+
+struct DesyncSpecialCase {
+    static const char *name() { return "DesyncSpecialCase"; }
+};
+
+std::vector<Q*> reader_queues(      ade::Graph &g,
+                                                 const ade::NodeHandle &obj) {
+    ade::TypedGraph<DataQueue> qgr(g);
+    std::vector<Q*> result;
+    for (auto &&out_eh : obj->outEdges()) {
+        result.push_back(qgr.metadata(out_eh).get<DataQueue>().q.get());
+    }
+    return result;
+}
+
+std::vector<Q*> input_queues(      ade::Graph &g,
+                                                const ade::NodeHandle &obj) {
+    ade::TypedGraph<DataQueue> qgr(g);
+    std::vector<Q*> result;
+    for (auto &&in_eh : obj->inEdges()) {
+        result.push_back(qgr.metadata(in_eh).contains<DataQueue>()
+                         ? qgr.metadata(in_eh).get<DataQueue>().q.get()
+                         : nullptr);
+    }
+    return result;
+}
+
+// FIXME: Is there a way to derive function from its GRunArgsP version?
+template<class C> using O = cv::util::optional<C>;
+void sync_data(cv::gimpl::stream::Result &r, cv::GOptRunArgsP &outputs) {
+    namespace own = cv::gapi::own;
+
+    for (auto && it : ade::util::zip(ade::util::toRange(outputs),
+                                     ade::util::toRange(r.args),
+                                     ade::util::toRange(r.flags))) {
+        auto &out_obj  = std::get<0>(it);
+        auto &res_obj  = std::get<1>(it);
+        bool available = std::get<2>(it);
+
+        using T = cv::GOptRunArgP;
+#define HANDLE_CASE(Type)                                               \
+        case T::index_of<O<Type>*>():                                   \
+            if (available) {                                            \
+                *cv::util::get<O<Type>*>(out_obj)                       \
+                    = cv::util::make_optional(std::move(cv::util::get<Type>(res_obj))); \
+            } else {                                                    \
+                cv::util::get<O<Type>*>(out_obj)->reset();              \
+            }
+
+        // FIXME: this conversion should be unified
+        switch (out_obj.index()) {
+            HANDLE_CASE(cv::Scalar);     break;
+            HANDLE_CASE(cv::RMat);       break;
+            HANDLE_CASE(cv::MediaFrame); break;
+
+        case T::index_of<O<cv::Mat>*>(): {
+            // Mat: special handling.
+            auto &mat_opt = *cv::util::get<O<cv::Mat>*>(out_obj);
+            if (available) {
+                auto q_map = cv::util::get<cv::RMat>(res_obj).access(cv::RMat::Access::R);
+                // FIXME: Copy! Maybe we could do some optimization for this case!
```

Comment:
Shouldn't we be concerned about this copy? Is there a task to fix it later if needed?